cpp/src/text/minhash.cu

```diff
@@ -51,79 +50,98 @@ namespace {
+template <typename HashFunction,
+          typename hash_value_type =
+            std::conditional_t<std::is_same_v<decltype(HashFunction::result_type), uint32_t>,
+                               uint32_t,
+                               uint64_t>>
+__global__ void minhash_fn(cudf::column_device_view const d_strings,
+                           cudf::device_span<hash_value_type const> seeds,
+                           cudf::size_type width,
+                           hash_value_type* d_hashes)
+{
+  auto const idx = static_cast<std::size_t>(threadIdx.x + blockIdx.x * blockDim.x);
+  if (idx >= (static_cast<std::size_t>(d_strings.size() *
+                                       static_cast<std::size_t>(cudf::detail::warp_size)))) {
+    return;
+  }
+
+  auto const str_idx  = static_cast<cudf::size_type>(idx / cudf::detail::warp_size);
+  auto const lane_idx = static_cast<cudf::size_type>(idx % cudf::detail::warp_size);
+
+  if (d_strings.is_null(str_idx)) { return; }
+
+  auto const d_str    = d_strings.element<cudf::string_view>(str_idx);
+  auto const d_output = d_hashes + (str_idx * seeds.size());
+
+  // initialize hashes output for this string
+  if (lane_idx == 0) {
+    auto const init = d_str.empty() ? 0 : std::numeric_limits<hash_value_type>::max();
+    thrust::fill(thrust::seq, d_output, d_output + seeds.size(), init);
+  }
+  __syncwarp();
+
+  auto const begin = d_str.data() + lane_idx;
+  auto const end   = d_str.data() + d_str.size_bytes();
+
+  // each lane hashes 'width'  substrings of d_str
+  for (auto itr = begin; itr < end; itr += cudf::detail::warp_size) {
+    if (cudf::strings::detail::is_utf8_continuation_char(*itr)) { continue; }
+    auto const check_str =  // used for counting 'width' characters
+      cudf::string_view(itr, static_cast<cudf::size_type>(thrust::distance(itr, end)));
+    auto const [bytes, left] = cudf::strings::detail::bytes_to_character_position(check_str, width);
+    if ((itr != d_str.data()) && (left > 0)) { continue; }  // true if past the end of the string
+
+    auto const hash_str = cudf::string_view(itr, bytes);
+    // hashing with each seed on the same section of the string is 10x faster than
+    // computing the substrings for each seed
+    for (std::size_t seed_idx = 0; seed_idx < seeds.size(); ++seed_idx) {
+      auto const hasher = HashFunction(seeds[seed_idx]);
+      // hash substring and store the min value
+      if constexpr (std::is_same_v<hash_value_type, uint32_t>) {
         auto const hvalue = hasher(hash_str);
-        cuda::atomic_ref<cudf::hash_value_type, cuda::thread_scope_block> ref{
-          *(d_output + seed_idx)};
+        cuda::atomic_ref<hash_value_type, cuda::thread_scope_block> ref{*(d_output + seed_idx)};
+        ref.fetch_min(hvalue, cuda::std::memory_order_relaxed);
+      } else {
+        auto const hvalue = thrust::get<0>(hasher(hash_str));  // just use the first uint64_t
```

Comment:
Can we add comments (or statically assert?) that this assumes `HashFunction` is `MurmurHash_x64_128`? The conditional checks the `hash_value_type` but that assumes behavior that wouldn't apply if we used XXHash (also comes in a 64 bit flavor but doesn't return a two-part value like MurmurHash_x64_128's 128 bit result).

Reply:
You would get a compile error either way. Adding another hash algorithm would be a big change so I don't think it would be accidentally dropped in here. Maybe I'm missing what you are trying to protect against.

Reply:
I agree an error would occur either way. I just want to make it explicit that this 64 bit minhash function is assuming the result layout of the 128 bit hash function.

Reply:
Would a comment suffice or are you looking for something else?

Reply:
Yup! Comments are fine.